File: csharp/LogicAndTrick.WikiCodeParser/Lines.cs
namespace LogicAndTrick.WikiCodeParser
{
    /// <summary>
    /// Line-based state tracker for a multi-line string
    /// </summary>
    public class Lines
    {
        public string[] Content { get; set; }
        public int Index { get; set; }

        public Lines(string content)
        {
            Content = content.Split('\n');
            Index = -1;
        }

        public void Back() => Index--;
        public bool Next() => ++Index < Content.Length;
        public string Value() => Content[Index];
        public int Current() => Index;
        public void SetCurrent(int index) => Index = index;
    }
}

File: javascript/src/Lines.ts

export class Lines {
    public Content : string[];
    public Index : number;

    constructor(content : string) {
        this.Content = content.split('\n');
        this.Index = -1;
    }

    public Back() {
        this.Index--;
    }
    public Next() : boolean {
        return ++this.Index < this.Content.length;
    }
    public Value() : string {
        return this.Content[this.Index];
    }
    public Current() : number {
        return this.Index;
    }
    public SetCurrent(index : number) {
        this.Index = index;
    }
}
